Add Cloudian HyperStore Object Storage

tpodowd opened the pull request:
### Description

This PR Adds A New Object Storage Provider Plugin for Cloudian HyperStore

- Allow the CloudStack administrator to connect to Cloudian HyperStore object storage.
- Once connected, CloudStack Accounts can create buckets that are managed by and belong to their own Account.
- IAM Credentials are available for each bucket such that Accounts can use the buckets either from 3rd party S3 applications or from the CloudStack Bucket Browser UI Feature.
- The plugin supports all the current CloudStack bucket operations such as Object Lock, Versioning, Encryption and policy settings.
- The plugin currently does not support setting a bucket quota as HyperStore does not currently support that functionality.
- Bucket usage is supported.

More Details:

- See plugins/storage/object/cloudian/README.md for details

UI Changes - Add Object Storage for Cloudian HyperStore:

- Cloudian HyperStore Object Storage requires more fields than Minio, Ceph and Simulator so when the Cloudian HyperStore provider is selected, the GUI adjusts and offers the extra fields that the provider requires.

Other Bug fixes and improvements as part of this fix I kept in separate commits.

- Use a password input field type for object store secret key entry
- Fix to avoid pre-pending a second '/' to the object name in the Bucket Browser Feature.
- Fix issue where usage may not be collected if another object store is down.
- Various fixes and enhancements to CloudianClient

<!--- ******************************************************************************* -->
<!--- NOTE: AUTOMATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ******************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):

<img width="589" alt="Screenshot 2024-09-30 at 13 02 20" src="https://github.com/user-attachments/assets/755d38e2-5c7c-4c06-a6ed-07d95aa08986">

### How Has This Been Tested?

1. Unit Testing
2. Testing against the Cloudian HyperStore system. Tested bad configurations and credentials, system down.
3. Monitored log entries and sniffed network connections.
4. Tested the various UI compontents such as Add Object Storage, Edit and Delete Object Storage. Adding buckets with various Accounts and Projects, editing bucket configurations, verifying configurations were set using different S3 exploration tools.
5. Tested the bucket browser UI component against HyperStore.

#### How did you try to break this feature and the system with this change?

- As this change altered CloudianClient which is shared code with the other cloudian infrastructure plugin, I also re-tested that plugin after these changes.
- I also lightly tested with Minio.
- This plugin feature shouldn't break other areas of the system.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11244

rohityadavcloud commented:
Thanks for the PR @tpodowd since we don't have the cloudian system to test against, we'll help with regression testing.

@blueorangutan test

tpodowd commented:
Hi @rohityadavcloud - I updated the PR to fix the README.md lint issue and also added a bit more unit test coverage in the main driver code. 

You mentioned the following:
> since we don't have the cloudian system to test against, we'll help with regression testing.

Thanks, let me know what information you need and I'll do my best to get back to you.

blueorangutan commented:
<b>[SF] Trillian test result (tid-11583)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 56540 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9748-t11583-kvm-ol8.zip
Smoke tests completed. 141 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

tpodowd commented:
Hi @DaanHoogland - I think I have got the pre-commit stuff nailed now.  I ran pre-commit locally and it fixed the end of files for me. Then I reviewed that and pushed commit 40c0366

tpodowd commented:
Sorry. I was reviewing the beautiful code coverage report and reviewing code I had not tested when I noticed a bad typo that means `CloudianClient` won't timeout. I have a fix locally and have added unit tests. I am doing a full build and a bit more testing and then I'll push another commit.

tpodowd commented:
Ok. Hopefully that is it. I have pre-commit hooked in now also so I know that is clean. Code coverage should be a little better again also.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11283

tpodowd commented:
Hi @DaanHoogland / @rohityadavcloud  - There seems to be an error in one of the checks. I'm not sure that it is related to my changes though. Let me know if I need to do anything about it. Thanks!

DaanHoogland commented:
> Hi @DaanHoogland / @rohityadavcloud - There seems to be an error in one of the checks. I'm not sure that it is related to my changes though. Let me know if I need to do anything about it. Thanks!

rerunning, let's see. It doesn't seem related to me either.

tpodowd commented:
I realised that I should not change the key names that the Object Store Details use as they may be read/updated outside of the plugin. Thanks!

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 11332

JoaoJandre commented on plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudianHyperStoreObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,877 @@
+                    if (iamAccessKeyId.equals(accessKeyMetadata.getAccessKeyId())) {
+                        return null;  // The IAM AccessKeyId still exists (as expected). return null.
+                    }
+                    // Usually, there will only be 1 credential that we manage, but an error persisting
+                    // credentials might leave an un-managed credential which we can just delete. It is better
+                    // to delete as otherwise, we may hit a max credential limit for this IAM user.
+                    DeleteAccessKeyRequest deleteAccessKeyRequest = new DeleteAccessKeyRequest();
+                    deleteAccessKeyRequest.setUserName(iamUser);
+                    deleteAccessKeyRequest.setAccessKeyId(accessKeyMetadata.getAccessKeyId());
+                    logger.info("Deleting un-managed IAM AccessKeyId {} for IAM User {}", accessKeyMetadata.getAccessKeyId(), iamUser);
+                    iamClient.deleteAccessKey(deleteAccessKeyRequest);
```
Is it really ok to delete the credential in this method? If some error created a lot of unmanaged credentials, shouldn't the admin delete them manually? (asking as a layman on Cloudian)

tpodowd replied:
Hi @JoaoJandre . Thanks for asking. Indeed it's a little strange at first but I think it is the right thing to do. This IAM user is owned by cloudstack. There should be nothing else creating IAM credentials for this user except for this plugin. IAM resources such as user/credentials/policies etc are the responsibility of the account so the administrator would not normally get involved with managing them. Also the account itself is owned by cloudstack and the idea is that you can only create buckets via the cloudstack interface and the credentials you have access to as a user also have no IAM access. That's basically why in this particular case, I think it is best to tidy up directly. Hope that answers your question.

JoaoJandre replied:
Ok, that makes sense then. In any case, this user deletion code is duplicated, we could extract it to a method.

tpodowd replied:
Done! Thanks for the suggestion.

JoaoJandre commented on plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudianHyperStoreObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,877 @@
+
+        if (! createdUser && iamAccessKeyId == null) {
+            // User already exists but we never saved any access key before. We should try clean up
+            logger.debug("Looking for any un-managed IAM credentials for IAM User {}", iamUser);
+            ListAccessKeysResult listRes = iamClient.listAccessKeys(new ListAccessKeysRequest().withUserName(iamUser));
+            for (AccessKeyMetadata accessKeyMetadata : listRes.getAccessKeyMetadata()) {
+                DeleteAccessKeyRequest deleteAccessKeyRequest = new DeleteAccessKeyRequest();
+                deleteAccessKeyRequest.setUserName(iamUser);
+                deleteAccessKeyRequest.setAccessKeyId(accessKeyMetadata.getAccessKeyId());
+                logger.info("Deleting un-managed IAM AccessKeyId {} for IAM User {}", accessKeyMetadata.getAccessKeyId(), iamUser);
+                iamClient.deleteAccessKey(deleteAccessKeyRequest);
+            }
```
Same here, should we be cleaning up stuff on a method used to create credentials?

tpodowd replied:
Thanks yes. Same answer really. The plugin should manage this for the account that is created specifically for CloudStack.

JoaoJandre commented on plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/util/CloudianHyperStoreUtil.java:
```diff
@@ -0,0 +1,218 @@
+    public static final String IAM_USER_USERNAME = "CloudStack";
+    public static final String IAM_USER_POLICY_NAME = "CloudStackPolicy";
+    public static final String IAM_USER_POLICY;
+    static {
+        StringBuilder sb = new StringBuilder();
+        sb.append("{\n");
+        sb.append("  \"Version\": \"2012-10-17\",\n");
+        sb.append("  \"Statement\": [\n");
+        sb.append("    {\n");
+        sb.append("      \"Sid\": \"AllowFullS3Access\",\n");
+        sb.append("      \"Effect\": \"Allow\",\n");
+        sb.append("      \"Action\": [\n");
+        sb.append("        \"s3:*\"\n");
+        sb.append("      ],\n");
+        sb.append("      \"Resource\": \"*\"\n");
+        sb.append("    },\n");
+        sb.append("    {\n");
+        sb.append("      \"Sid\": \"ExceptBucketCreationOrDeletion\",\n");
+        sb.append("      \"Effect\": \"Deny\",\n");
+        sb.append("      \"Action\": [\n");
+        sb.append("        \"s3:createBucket\",\n");
+        sb.append("        \"s3:deleteBucket\"\n");
+        sb.append("      ],\n");
+        sb.append("      \"Resource\": \"*\"\n");
+        sb.append("    }\n");
+        sb.append("  ]\n");
+        sb.append("}\n");
+        IAM_USER_POLICY = sb.toString();
+    }
```
Why not simply define the variable with the plain string value? Is there any reason to use StringBuilder here?

tpodowd replied:
There is no strong reason except that I was frustrated there was no """ in Java11 :-) I can rework it into a standard string and remove the static initialization I guess.

JoaoJandre commented on plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/util/CloudianHyperStoreUtil.java:
```diff
@@ -0,0 +1,218 @@
+                port = DEFAULT_ADMIN_PORT;
+            }
+            return new CloudianClient(host, port, scheme, user, pass, validateSSL, getAdminTimeoutSeconds());
+        } catch (MalformedURLException e) {
+            throw new CloudRuntimeException(e);
+        } catch (KeyStoreException | NoSuchAlgorithmException | KeyManagementException e) {
+            throw new CloudRuntimeException(e);
+        }
```
We could join these catches

tpodowd replied:
ok. will do.

tpodowd commented on plugins/integrations/cloudian/src/main/java/org/apache/cloudstack/cloudian/client/CloudianClient.java:
```diff
@@ -159,22 +180,108 @@ private HttpResponse post(final String path, final Object item) throws IOExcepti
+            cmd.append(userId);
+        }
+        if (! StringUtils.isBlank(bucket)) {
+            // Assume userId is also set (or request fails).
```
Thanks for your time @JoaoJandre. Can't argue with this. I will fix it.

tpodowd commented on plugins/integrations/cloudian/src/main/java/org/apache/cloudstack/cloudian/client/CloudianClient.java:
```diff
@@ -287,33 +492,42 @@ public CloudianGroup listGroup(final String groupId) {
+            if (noResponseEntity(response)) {
+                throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, "API error");
+            }
+            // The empty list case is badly behaved and returns as 200 with an empty body.
+            // We'll try detect this by reading the first byte and then putting it back if required.
+            PushbackInputStream iStream = new PushbackInputStream(response.getEntity().getContent());
+            int firstByte=iStream.read();
+            if (firstByte == -1) {
+                return new ArrayList<>(); // EOF => empty list
             }
+            // unread that first byte and process the JSON
+            iStream.unread(firstByte);
```
Sure. I'll do that also.

tpodowd commented on plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudianHyperStoreObjectStoreDriverImpl.java:
```diff
@@ -0,0 +1,877 @@
+import com.cloud.utils.exception.CloudRuntimeException;
+
+public class CloudianHyperStoreObjectStoreDriverImpl extends BaseObjectStoreDriverImpl {
+    protected Logger logger = LogManager.getLogger(CloudianHyperStoreObjectStoreDriverImpl.class);
```
Ah ok. Will fix this also. Thanks!

tpodowd commented:
Hi @JoaoJandre - I have pushed another commit to address your review comments. Thanks for your time and let me know if you have any other concerns or questions.

JoaoJandre commented:
> Hi @JoaoJandre - I have pushed another commit to address your review comments. Thanks for your time and let me know if you have any other concerns or questions.

@tpodowd I did my best with no knowledge of Cloudian :smile: . In any case, overall it looks good to me. There is only one thing (https://github.com/apache/cloudstack/pull/9748#discussion_r1806353907) left that I think should be addressed.

tpodowd commented:
> @tpodowd I did my best with no knowledge of Cloudian 😄 . In any case, overall it looks good to me. There is only one thing ([#9748 (comment)](https://github.com/apache/cloudstack/pull/9748#discussion_r1806353907)) left that I think should be addressed.

Hi @JoaoJandre - No worries. Thank you so much again for your time on this.  I have addressed your last comment  and have pushed an update.

JoaoJandre reviewed (approved):
LGTM, did not test it.

DaanHoogland commented on plugins/storage/object/cloudian/add_cloudian_hyperstore.png:
why would we need this image in the production source? seems more for the documentation.

DaanHoogland commented on plugins/storage/object/cloudian/README.md:
should this be added to a PR for https://github.com/apache/cloudstack-documentation?

tpodowd replied:
Hi @DaanHoogland. I can certainly move the the `png` out. Is the README file ok (once I remove the `png` reference from it)? I was trying to leave a little developer documentation on the plugin. I think we need user documentation separately as you say and the `png` can go there I guess.

DaanHoogland replied:
fine, but keep in mind the docs are better viewed than any readme in the code (unless it is really *in* the code)

tpodowd replied:
Ok. Thanks. I updated the README and deleted the `png` file.

DaanHoogland reviewed (commented):
clgtm, one organisational issue though

abh1sar commented:
Hi @tpodowd,
I understand Cloudian doesn't support per bucket Quota.
Does it support per user quota? Is there a way to configure it from within CloudStack?

cc @rohityadavcloud @sureshanaparti

tpodowd commented:
Hi @abh1sar - Thanks for your comment/question.

> I understand Cloudian doesn't support per bucket Quota.

Yes, Cloudian HyperStore does not currently support a bucket storage quota.

>  Does it support per user quota?

Yes, Cloudian HyperStore does support per user quota. We have a warning level and a hard limit for storage bytes and we also have some other related settings.

> Is there a way to configure it from within CloudStack?

Unfortunately not. There are some issues here:

1. Although the CloudStack APIs for Object Storage allow setting a quota on a bucket, there is no API framework provided to the plugins to set something on the CloudStack Account level.
2. On HyperStore, setting a quota on an Account/User is usually something that an administrator would do as it is related to QoS settings (ie protecting the system, rather than protecting the user). I guess a user concerned about potential storage costs might also want to impose a limit on themselves. But, if the administrator for example set the QoS limit to one thing, then the user should not be able to raise it higher or disable it for themselves.

Currently, the administrator would have to login to the HyperStore system and either:
1. Select the HyperStore group (representing the CloudStack Domain) and set a QoS limit that applies to all users in that group.
2. Select the HyperStore user (representing the CloudStack Account) and set a QoS limit that only applies to that particular user.

abh1sar commented:
Thanks @tpodowd for your response.
If we had to implement the functionality to set Account level Quota in cloudstack, how do you think that could be done?
Does the HyperStore plugin has API to do something like that?
I am asking because I am working on a PR which adds resource limits to Object Storage space usage.

tpodowd commented:
Hi @abh1sar. The current plugin does not have any API support itself for setting Account level QoS (as it only implements the provided plugin APIs). The HyperStore Admin API itself does support setting QoS for a user though so that could be made available to CloudStack if it was implemented in CloudianClient (which is easy enough). If I know more about what you are doing, I guess I can also chip in. 

FYI. I have another PR pending this one also which adds the ability to edit the Object Store details.
https://github.com/tpodowd/cloudstack/tree/edit_object_storage

tpodowd commented:
Hi @DaanHoogland @rohityadavcloud - Any update on when this PR is likely to be merged?

abh1sar commented:
Hi @tpodowd ,
Please sync with the latest code and test bucket creation. The quota field in create bucket api has been made mandatory since https://github.com/apache/cloudstack/pull/10017 to support Object Storage Limits per Account/Project and Domain.
Happy to discuss further if needed.

tpodowd commented:
Hi @abh1sar, ok. I will allocate some time to look at this in the coming days. As I mentioned before, we don't have a bucket level quota setting. We do have a User level QoS setting. I think the only thing that I can do currently is to set the User (that owns the bucket) QoS settings to the limit specified in the create bucket request. An example might be as below:

- Bucket B1 created for User A with a quota of 5GB.
- > Plugin adjusts the HyperStore user A's quota to 5GB
- Bucket B2 created for User A with a quota of 5GB.
- > Plugin keeps HyperStore user A's quota at 5GB (updates CS DB quotas for B1,B2 to 5GB to match)
- Bucket B3 created for User A with a quota of 3GB.
- > Plugin reduces HyperStore user A's quota to 3GB (updates CS DB quotas for B1,B2,B3 to 3GB to match)
- Bucket B2's quota is adjusted to 10GB
- > Plugin expands HyperStore user A's quota to 10GB (updates CS DB quotas for B1,B2,B3 to 10GB to match)

When any adjustment happens to the user's quota, all of the HyperStore buckets owned by that account need to get their quota settings updated in the DB to correctly represent the latest user's quota setting.

This is currently all we can do I think. If we document it, it should not be too hard to understand for the admin.

Does this sounds ok? I think plugin wise, the plugin currently does not update information about different bucket(s) when processing a request for a certain bucket, so I will need to figure out how best to do that. I do think it is easier for the administrator to understand the implementation and understand what the quota is set to though if it is done this way.

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9 ✖️ debian ✖️ suse15. SL-JID 12496

DaanHoogland commented:
@tpodowd , it seems there is some dependency error on debian: 
```
09:14:14 [ERROR] [ERROR] Some problems were encountered while processing the POMs:
09:14:14 [FATAL] Non-resolvable parent POM for org.apache.cloudstack:cloud-plugin-storage-object-cloudian:4.20.0.0-SNAPSHOT: Could not find artifact org.apache.cloudstack:cloudstack-plugins:pom:4.20.0.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 24, column 13
```
and in the github action:
```
Error: ] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.apache.cloudstack:cloud-plugin-storage-object-cloudian:4.20.0.0-SNAPSHOT: The following artifacts could not be resolved: org.apache.cloudstack:cloudstack-plugins:pom:4.20.0.0-SNAPSHOT (absent): Could not find artifact org.apache.cloudstack:cloudstack-plugins:pom:4.20.0.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 24, column 13
```
I must say I can not find it (yet)

shwstppr commented on plugins/storage/object/cloudian/pom.xml:
```diff
@@ -0,0 +1,70 @@
+    <parent>
+        <groupId>org.apache.cloudstack</groupId>
+        <artifactId>cloudstack-plugins</artifactId>
+        <version>4.20.0.0-SNAPSHOT</version>
```
```suggestion
        <version>4.21.0.0-SNAPSHOT</version>
```

tpodowd replied:
This is resolved. Thanks.

tpodowd commented:
Hi @DaanHoogland - please ignore this PR for the moment. I'm going to look at the plugin again with the new information that  @abh1sar has provided regarding the quota being required. I'll fix the build issue before I push anything to the branch.

@abh1sar - can you check my reply above also regarding how I think we might handle this? Thanks @shwstppr for pointing out the issue with the pom versions.

tpodowd commented:
I got my local build working again and confirmed that indeed we cannot add buckets currently as quota is now a requirement and create bucket fails with:
> Failed to create bucket with name: testbucket com.cloud.utils.exception.CloudRuntimeException: This bucket does not support quotas.

This is because the Cloudian plugin as written does not implement bucket quota and fails the request. I need to figure out an approach for this. I guess there are two approaches:
1. The plugin ignores the quota and creates the bucket regardless.
2. We implement the user quota that I mentioned in an earlier comment.

I am guessing that option 1 is not really an option and I need to implement option 2? Let me know if I am wrong about that.

abh1sar commented:
> Hi @abh1sar, ok. I will allocate some time to look at this in the coming days. As I mentioned before, we don't have a bucket level quota setting. We do have a User level QoS setting. I think the only thing that I can do currently is to set the User (that owns the bucket) QoS settings to the limit specified in the create bucket request. An example might be as below:
> 
> * Bucket B1 created for User A with a quota of 5GB.
> * > Plugin adjusts the HyperStore user A's quota to 5GB
> * Bucket B2 created for User A with a quota of 5GB.
> * > Plugin keeps HyperStore user A's quota at 5GB (updates CS DB quotas for B1,B2 to 5GB to match)
> * Bucket B3 created for User A with a quota of 3GB.
> * > Plugin reduces HyperStore user A's quota to 3GB (updates CS DB quotas for B1,B2,B3 to 3GB to match)
> * Bucket B2's quota is adjusted to 10GB
> * > Plugin expands HyperStore user A's quota to 10GB (updates CS DB quotas for B1,B2,B3 to 10GB to match)
> 
> When any adjustment happens to the user's quota, all of the HyperStore buckets owned by that account need to get their quota settings updated in the DB to correctly represent the latest user's quota setting.
> 
> This is currently all we can do I think. If we document it, it should not be too hard to understand for the admin.
> 
> Does this sounds ok? I think plugin wise, the plugin currently does not update information about different bucket(s) when processing a request for a certain bucket, so I will need to figure out how best to do that. I do think it is easier for the administrator to understand the implementation and understand what the quota is set to though if it is done this way.

Hi @tpodowd
There could be another approach : if the user has 3 buckets of size 5GB, 10GB and 3GB, we keep the quota in CS DB as 5, 10 and 3 respectively. On the plugin side we can set the total user quota to be the sum of all quotas i.e. 18GB. So, the plugin needs to increase, decrease user quota whenever a bucket is created/ destroyed or quota is changed. This will also
need to be documented, that a bucket from the same user can feed space from another bucket of the same user (Although we defined bucket quota as 5, 10 and 3, the actually usage could be 6, 6 and 6.)

Another thing could be, if you are ok with Cloudian not supporting Object Storage Limits, you can document that and make the quota field non-mandatory for Cloudian.

tpodowd commented:
We discussed this internally. HyperStore does not currently support Bucket Quota limits. We have added an internal feature request for this which we may implement in a future HyperStore release. At such time, we will re-address and update the CloudStack HyperStore plugin bucket quota functionality. As the CloudStack UI and API currently both require a quota setting, we decided to allow the user to specify a value of 0 as the quota. The 0 quota will indicate no quota restriction for the bucket.

I rebased this branch on top of the latest main branch. 
I fixed 2 issues in commit 01be768a6955313f6e5962b6939124a97e54481c:
1. The pom.xml was updated for the correct version number.
2. Quota of 0 is now accepted and I added an extra test for the 0 case.

I built everything cleanly and tested adding a bucket successfully using a quota of 0. I also confirmed the error case worked with any other quota value and that UI shows the user the exception message to help them choose the 0 value.

Please let me know if you have any questions or comments.

DaanHoogland commented:
@blueorangutan package

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✖️ debian ✔️ suse15. SL-JID 12577

abh1sar commented:
> We discussed this internally. HyperStore does not currently support Bucket Quota limits. We have added an internal feature request for this which we may implement in a future HyperStore release. At such time, we will re-address and update the CloudStack HyperStore plugin bucket quota functionality. As the CloudStack UI and API currently both require a quota setting, we decided to allow the user to specify a value of 0 as the quota. The 0 quota will indicate no quota restriction for the bucket.
> 
> I rebased this branch on top of the latest main branch. I fixed 2 issues in commit [01be768](https://github.com/apache/cloudstack/commit/01be768a6955313f6e5962b6939124a97e54481c):
> 
> 1. The pom.xml was updated for the correct version number.
> 2. Quota of 0 is now accepted and I added an extra test for the 0 case.
> 
> I built everything cleanly and tested adding a bucket successfully using a quota of 0. I also confirmed the error case worked with any other quota value and that UI shows the user the exception message to help them choose the 0 value.
> 
> Please let me know if you have any questions or comments.

LGTM

DaanHoogland commented:
> Please let me know if you have any questions or comments.

thanks @tpodowd , this means you are done with this PR, for now? (so we run regression tests and merge ..)

tpodowd commented:
Hi @DaanHoogland - yes. I am done with this particular PR so it would be great to move forward. Thanks!

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 12598

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (ol8 mgmt + kvm-ol8) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-12506)</b>
Environment: kvm-ol8 (x2), Advanced Networking with Mgmt server ol8
Total time taken: 57047 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9748-t12506-kvm-ol8.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_06_purge_expunged_vm_background_task | `Failure` | 383.31 | test_purge_expunged_vms.py

tpodowd commented:
Hi @DaanHoogland - I saw you closed this and re-opened it but since it is not merged yet I was wondering if there is anything holding it up? I've another PR waiting for this one to be merged which adds editing of the Object Storage settings.

DaanHoogland commented:
@tpodowd 

> Hi @DaanHoogland - I saw you closed this and re-opened it but since it is not merged yet I was wondering if there is anything holding it up? I've another PR waiting for this one to be merged which adds editing of the Object Storage settings.

No, except maybe a 3rd party test report. But as this is an integration of 3rd party components we are lenient in that perspect, so I'm merging. Usually we ask that somene else than the contributer has tested and fed back their findings.

tpodowd commented:
Thanks for the explanation and for merging @DaanHoogland.
